Add IECoreScene::ShaderNetwork class

johnhaddon opened the pull request:
Until now, we've represented shader networks in Cortex/Gaffer using attributes containing `IECore.ObjectVector`, and represented connections between shaders using parameter values containing string values of the form "link:\<someHandle\>" where the shader associated with "\<someHandle\>" is identified by a fake parameter called "__handle". This was done for expedience a long time ago, but is really unsatisfactory for doing any serious manipulation of networks. So, using https://github.com/GafferHQ/gaffer/issues/2839 as an excuse I'm introducing a new ShaderNetwork class which is tailored to the task at hand.

This PR includes everything needed for a first pass implementation of shader tweaks, but I anticipate the following additions fairly soon :

- A method whereby we can represent "promoted" parameters from a Gaffer node network. I'm anticipating connections to these will be represented as `Connection( { "", "promotedParameter" }, { "internalNode", "internalParameter" } )` - i.e. just a regular connection, but with an empty shader as the source.
- A method of attaching metadata to the shaders/parameters so we can improve the downstream representation of shading networks when browsing them in the ShaderTweaks/SceneInspector UIs. I'm undecided on how to do this - it'll either be blind data, some new API, or some combination of the two.

danieldresser-ie commented:
I haven't looked in real detail yet, but looks pretty good to me.  I do still feel that we could use a few more warnings about the requirements we're placing on usage for performance reasons in ShaderNetwork.h, about ShaderNetwork taking ownership of what you pass, and how if you use getShader to modify a shader, you must complete your edits before doing anything that would trigger hash() to be called.

johnhaddon commented on include/IECoreScene/ShaderNetworkAlgo.h:
```diff
@@ -0,0 +1,66 @@
+IECORESCENE_API void removeUnusedShaders( ShaderNetwork *network );
+
+/// Performs a depth-first traversal of the network from the starting shader, calling
+/// `visitor` exactly once for each shader encountered.
```
Done in 574ee0b.

johnhaddon commented:
> I do still feel that we could use a few more warnings about the requirements we're placing on usage for performance reasons in ShaderNetwork.h, about ShaderNetwork taking ownership of what you pass, and how if you use getShader to modify a shader, you must complete your edits before doing anything that would trigger hash() to be called.

Yeah, sorry, I said I'd do this before making a PR didn't I? Added in 3c17a7b.

I'm now wondering if this approach is the right one though. I think for the use cases we care about we might be able to get better performance by taking a more watertight approach. Our two main use cases are generating a ShaderNetwork from a node network in Gaffer, and then modifying it downstream in a ShaderTweaks node.

Let's look at ShaderTweaks first. this receives a `const ShaderNetwork`, copies it, modifies the copy, and returns it. Because ShaderNetwork doesn't have unique ownership of the shaders, the `ShaderNetwork::copy()` call must do a full deep copy of everything. We can then modify any given shader in place at will, subject to the `hash()` restrictions we've just documented. But typically the network will be large, and we'll only be editing one or two shaders. If instead the ShaderNetwork had full ownership of the shaders, and only provided const access to them, `ShaderNetwork::copy()` could do a shallow copy. ShaderTweaks would then do a const `network->getShader()`, copy the result, modify it, and insert it back with `setShader()`. So we'd only be paying for copies of the things we modified, and the `hash()` caching would be watertight. We'd also be able to cache the hash of each individual shader if we wanted, speeding up subsequent rehashes.

Next, lets look at ShaderNetwork construction. To guarantee full ownership, `addShader()` would have to copy the shader passed to it. So a naive implementation would pay for a copy of every shader at the generation stage - a net loss if there are no downstream tweaks. But we could have an explicit mechanism for saying "here, you take it without copying, I promise I'm the only owner and I'm not going to change it". That might be better than the less explicit optimisations we have now. We might even be able to enforce the restriction with `addShader( std::move( myShaderPtr ) )` - this would check that the reference count was 1 (no other owners) and then "steal" the reference, nullifying `myShaderPtr`, thus heavily discouraging subsequent modification.

What do you think? I'm not 100% sure the performance matters either way (we know we need the cached hash, but don't yet know how much copying will hurt because we haven't got ShaderTweaks yet), but I'm interested enough that I'd like to sketch out an implementation this way and see how it feels.

johnhaddon commented:
I've pushed 47f0561, which contains the watertight hash implementation we spoke about, plus some use of move semantics to avoid copies. In theory I believe this gives us the best combination of robustness and performance, but I've yet to update the client code in Gaffer. Once I've done that, I'll have a better idea how I feel about this...

johnhaddon commented:
OK, see https://github.com/GafferHQ/gaffer/pull/2902/commits/c5590e8bae9922332093affee1a5ab1d5b15b699 for the Gaffer-side changes corresponding to the latest commits. It was pretty straightforward, but with one annoyance. Because the Shader node wants to generate parameter values and connections at the same, we have a bit of an ordering issue - we can't get the shader handle until we've added the shader to the network, but we can't add the shader to the network until we have all the parameters (and we need the handle to generate the connections). I've dealt with it easily enough so I don't think it's a big deal if you prefer the new API - I've also been considering separating the parameter value generation from the connection generation anyway...

johnhaddon commented:
I've pushed a series of commits that allow us to emulate component-level connections in OSL. The idea is that the user will author networks as they want to, making direct connections between the rgb/xyz components of colors/vectors, and we post-process the network to insert conversion shaders at the last minute when rendering. Hopefully this shouldn't be needed for long, as OSL 1.10 brings native support. New commits start at 288b922.

I'd very much like to get this reviewed/merged soon so I can continue progress on the Gaffer side...

johnhaddon commented on src/IECoreScene/ShaderNetwork.cpp:
```diff
@@ -0,0 +1,764 @@
+			auto it = m_nodes.find( handle );
+			if( it == m_nodes.end() )
+			{
+				throw IECore::Exception( boost::str(
```
Given the size of the networks we use in production I would prefer the more succinct form I've used already.

johnhaddon commented on test/IECoreScene/ShaderNetworkTest.py:
```diff
@@ -0,0 +1,504 @@
+		self.assertEqual( n.outputShader(), None )
+
+		n.setOutput( n.Parameter( "s1", "" ) )
+		self.assertEqual( n.getOutput(), n.Parameter( "s1", "" ) )
```
I did it deliberately to avoid verbosity. They both refer to the exact same thing, except the latter is actually the even more verbose `IECoreScene.ShaderNetwork.Parameter()`. If you think it's misleading I can replace them all with the long form, or I can use the implicit tuple->Parameter conversion I added later as a convenience.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -105,3 +110,128 @@ void ShaderNetworkAlgo::removeUnusedShaders( ShaderNetwork *network )
+const InternedString g_inParameterName( "in" );
+const InternedString g_outParameterName( "out" );
+const InternedString g_packInParameterNames[3] = { "in1", "in2", "in3" };
+const boost::regex g_componentRegex( "^(.*)\\.([rgbxyz])$" );
```
Because we're stuck needing to support GCC 4.8 for as long as we're either on legacy versions of certain DCCs or the current versions of the same DCCs don't adhere to VFXPlatform. And GCC 4.8's regex support is [somewhat lacking](https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions). Here's one particularly illuminating quote from that link :

> Seriously though, who though that shipping an implementation of regex_search that only does "return false" was a good idea?

johnhaddon commented on include/IECoreScene/ShaderNetwork.h:
```diff
@@ -0,0 +1,214 @@
+#define IECORESCENE_SHADERNETWORK_H
+
+#include "IECoreScene/TypeIds.h"
+#include "IECoreScene/Shader.h"
```
Fixed in 1ad138af87770d5f7f3638bb36fc50a6407cc1e5.

johnhaddon commented on include/IECoreScene/ShaderNetworkAlgo.h:
```diff
@@ -0,0 +1,72 @@
+
+/// Adds all shaders from `sourceNetwork` into `network`, renaming their handles as necessary to
+/// preserve uniqueness. Returns the parameter within `destinationNetwork` that corresponds to `sourceNetwork->getOutput()`.
+IECORESCENE_API ShaderNetwork::Parameter addShaders( ShaderNetwork *network, const ShaderNetwork *sourceNetwork, bool connections = true );
```
Fixed in 1ad138af87770d5f7f3638bb36fc50a6407cc1e5. I don't think we should require documentation for things as obvious as this one, but I realised I'd failed to document the less obvious signature for the `Visitor` class below, so included both fixes together.

johnhaddon commented on include/IECoreScene/ShaderNetwork.inl:
```diff
@@ -0,0 +1,188 @@
+			return *m_connection;
+		}
+
+		const void *m_container;
```
Done in 1ad138a.

johnhaddon commented on test/IECoreScene/ShaderNetworkTest.py:
```diff
@@ -0,0 +1,504 @@
+
+	def testHashRepeatability( self ) :
+
+		IECoreScene.ShaderNetwork( shaders = { "flat" : IECoreScene.Shader( "flat" ) } )
```
Nope - fixed in 32c5928.

johnhaddon commented:
Just pushed two additional commits to fix bugs I found this afternoon...

johnhaddon commented:
Any further comments? If not, I'd like to get this merged, since it's blocking progress on the Gaffer side...
